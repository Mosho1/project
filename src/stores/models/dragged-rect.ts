import { types, hasParent, getParent } from 'mobx-state-tree';
import { IBox } from './box';
import { IStore, IStage } from '../domain-state';

export const DraggedRect = types.model('DraggedRect', {
    startX: types.number,
    startY: types.number,
    endX: types.number,
    endY: types.number,
}).views(self => ({
    get store(): null | IStore {
        if (!hasParent(self, 1)) return null;
        return getParent(self, 1);
    },
    get width() {
        return self.endX - self.startX;
    },
    get height() {
        return self.endY - self.startY;
    },
})).views(self => ({
    get stage(): null | IStage {
        if (!self.store) return null;
        return self.store.stage;
    }
})).views(self => ({
    get absoluteCoords(): { x: number, y: number } {
        if (!self.stage) return { x: self.startX, y: self.startY };
        return self.stage.getAbsolutePosition(
            self.startX,
            self.startY
        );
    }
})).views(self => ({
    get x() {
        return self.width > 0 ? self.absoluteCoords.x : self.absoluteCoords.x + self.width;
    },
    get y() {
        return self.height > 0 ? self.absoluteCoords.y : self.absoluteCoords.y + self.height;
    },
})).views(self => ({
    hasIntersection({ x, y, width, height }: IBox) {
        return !(self.x > x + width ||
            self.x + Math.abs(self.width) < x ||
            self.y > y + height ||
            self.y + Math.abs(self.height) < y);
    }
})).actions(self => {
    const start = (x: number, y: number) => {
        self.startX += x;
        self.startY += y;
        return self;
    };

    const end = (x: number, y: number) => {
        self.endX += x;
        self.endY += y;
        return self;
    };

    return {
        start, end
    };
});

type IDraggedRectType = typeof DraggedRect.Type;
export interface IDraggedRect extends IDraggedRectType { };