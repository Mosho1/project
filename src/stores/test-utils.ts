import { mapValues } from 'lodash';

export const mock = <T>(instance: T, mock: NestedPartial<T>) => {
    for (let k in mock) {
        const descriptor = Object.getOwnPropertyDescriptor(instance, k);
        Object.defineProperties(instance, mapValues(mock, value => ({
            value,
            enumerable: descriptor!.enumerable,
            configurable: true,
            writable: true,
        })));
    }
    return instance;
};

const f = (a: any[], b: any[]): any[] =>
    [].concat(...a.map(a2 => b.map((b2: any) => [].concat(a2 as any, b2 as any) as any) as any));

export const product = (a: any[], b: any[], ...c: any[]): any[][] => {
    if (!b || b.length === 0) {
        return a;
    }
    const [b2, ...c2] = c;
    const fab = f(a, b);
    return product(fab, b2, ...c2);
};