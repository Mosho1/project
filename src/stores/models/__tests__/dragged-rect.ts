import { DraggedRect, IDraggedRect } from "../dragged-rect";
import { product, mock } from '../../test-utils';
import { range } from 'lodash';

let a: IDraggedRect
beforeEach(() => {
    a = DraggedRect.create({ startX: 0, startY: 0, endX: 50, endY: 50 });
});

test('start/end/width/height', () => {
    const moves = product(range(-100, 100, 40), range(-100, 100, 40));
    for (const [i, [x, y]] of moves.entries()) {
        (i % 2 === 0 ? a.start : a.end)(x, y);
        const snapshotName = `${x},${y}`;
        expect(a).toMatchSnapshot(snapshotName);
        expect(a).toMatchSnapshot(snapshotName);
        expect(a.x).toMatchSnapshot(snapshotName);
        expect(a.y).toMatchSnapshot(snapshotName);
        expect(a.width).toMatchSnapshot(snapshotName);
        expect(a.height).toMatchSnapshot(snapshotName);
    }
});

test('absoluteCoords', () => {
    expect(a.absoluteCoords).toEqual({ x: 0, y: 0 });
    a.start(10, 10);
    expect(a.absoluteCoords).toEqual({ x: 10, y: 10 });
    let stage = { getAbsolutePosition: (x: any) => x } as any;
    spyOn(stage, 'getAbsolutePosition');
    mock(a, { stage });
    a.absoluteCoords;
    expect(stage.getAbsolutePosition).toHaveBeenCalledWith(10, 10);
    a.start(10, 10);
    a.absoluteCoords;
    expect(stage.getAbsolutePosition).toHaveBeenCalledWith(20, 20);
});

test('hasIntersection', () => {
    const rects = product(
        range(-50, 51, 50),
        range(-50, 51, 50),
        [50],
        [50]
    );

    const boxes = product(
        range(-150, 151, 150),
        range(-150, 151, 150),
        [100],
        [100]
    );

    for (const [a, b, c, d] of rects) {
        const rect = DraggedRect.create({ startX: a, startY: b, endX: c, endY: d });
        for (const [x, y, width, height] of boxes) {
            expect(rect.hasIntersection({ x, y, width, height } as any)).toMatchSnapshot()
        }
    }

});