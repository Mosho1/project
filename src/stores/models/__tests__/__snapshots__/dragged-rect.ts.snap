// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hasIntersection 1`] = `true`;

exports[`hasIntersection 2`] = `true`;

exports[`hasIntersection 3`] = `false`;

exports[`hasIntersection 4`] = `true`;

exports[`hasIntersection 5`] = `true`;

exports[`hasIntersection 6`] = `false`;

exports[`hasIntersection 7`] = `false`;

exports[`hasIntersection 8`] = `false`;

exports[`hasIntersection 9`] = `false`;

exports[`hasIntersection 10`] = `false`;

exports[`hasIntersection 11`] = `true`;

exports[`hasIntersection 12`] = `false`;

exports[`hasIntersection 13`] = `false`;

exports[`hasIntersection 14`] = `true`;

exports[`hasIntersection 15`] = `false`;

exports[`hasIntersection 16`] = `false`;

exports[`hasIntersection 17`] = `false`;

exports[`hasIntersection 18`] = `false`;

exports[`hasIntersection 19`] = `false`;

exports[`hasIntersection 20`] = `true`;

exports[`hasIntersection 21`] = `false`;

exports[`hasIntersection 22`] = `false`;

exports[`hasIntersection 23`] = `true`;

exports[`hasIntersection 24`] = `false`;

exports[`hasIntersection 25`] = `false`;

exports[`hasIntersection 26`] = `false`;

exports[`hasIntersection 27`] = `false`;

exports[`hasIntersection 28`] = `false`;

exports[`hasIntersection 29`] = `false`;

exports[`hasIntersection 30`] = `false`;

exports[`hasIntersection 31`] = `true`;

exports[`hasIntersection 32`] = `true`;

exports[`hasIntersection 33`] = `false`;

exports[`hasIntersection 34`] = `false`;

exports[`hasIntersection 35`] = `false`;

exports[`hasIntersection 36`] = `false`;

exports[`hasIntersection 37`] = `false`;

exports[`hasIntersection 38`] = `false`;

exports[`hasIntersection 39`] = `false`;

exports[`hasIntersection 40`] = `false`;

exports[`hasIntersection 41`] = `true`;

exports[`hasIntersection 42`] = `false`;

exports[`hasIntersection 43`] = `false`;

exports[`hasIntersection 44`] = `false`;

exports[`hasIntersection 45`] = `false`;

exports[`hasIntersection 46`] = `false`;

exports[`hasIntersection 47`] = `false`;

exports[`hasIntersection 48`] = `false`;

exports[`hasIntersection 49`] = `false`;

exports[`hasIntersection 50`] = `true`;

exports[`hasIntersection 51`] = `false`;

exports[`hasIntersection 52`] = `false`;

exports[`hasIntersection 53`] = `false`;

exports[`hasIntersection 54`] = `false`;

exports[`hasIntersection 55`] = `false`;

exports[`hasIntersection 56`] = `false`;

exports[`hasIntersection 57`] = `false`;

exports[`hasIntersection 58`] = `true`;

exports[`hasIntersection 59`] = `true`;

exports[`hasIntersection 60`] = `false`;

exports[`hasIntersection 61`] = `false`;

exports[`hasIntersection 62`] = `false`;

exports[`hasIntersection 63`] = `false`;

exports[`hasIntersection 64`] = `false`;

exports[`hasIntersection 65`] = `false`;

exports[`hasIntersection 66`] = `false`;

exports[`hasIntersection 67`] = `false`;

exports[`hasIntersection 68`] = `true`;

exports[`hasIntersection 69`] = `false`;

exports[`hasIntersection 70`] = `false`;

exports[`hasIntersection 71`] = `false`;

exports[`hasIntersection 72`] = `false`;

exports[`hasIntersection 73`] = `false`;

exports[`hasIntersection 74`] = `false`;

exports[`hasIntersection 75`] = `false`;

exports[`hasIntersection 76`] = `false`;

exports[`hasIntersection 77`] = `true`;

exports[`hasIntersection 78`] = `false`;

exports[`hasIntersection 79`] = `false`;

exports[`hasIntersection 80`] = `false`;

exports[`hasIntersection 81`] = `false`;

exports[`start/end/width/height: -20,-20 1`] = `
Object {
  "endX": -350,
  "endY": -110,
  "startX": -460,
  "startY": -220,
}
`;

exports[`start/end/width/height: -20,-20 2`] = `
Object {
  "endX": -350,
  "endY": -110,
  "startX": -460,
  "startY": -220,
}
`;

exports[`start/end/width/height: -20,-20 3`] = `-460`;

exports[`start/end/width/height: -20,-20 4`] = `-220`;

exports[`start/end/width/height: -20,-20 5`] = `110`;

exports[`start/end/width/height: -20,-20 6`] = `110`;

exports[`start/end/width/height: -20,-60 1`] = `
Object {
  "endX": -350,
  "endY": -110,
  "startX": -440,
  "startY": -200,
}
`;

exports[`start/end/width/height: -20,-60 2`] = `
Object {
  "endX": -350,
  "endY": -110,
  "startX": -440,
  "startY": -200,
}
`;

exports[`start/end/width/height: -20,-60 3`] = `-440`;

exports[`start/end/width/height: -20,-60 4`] = `-200`;

exports[`start/end/width/height: -20,-60 5`] = `90`;

exports[`start/end/width/height: -20,-60 6`] = `90`;

exports[`start/end/width/height: -20,-100 1`] = `
Object {
  "endX": -330,
  "endY": -50,
  "startX": -440,
  "startY": -200,
}
`;

exports[`start/end/width/height: -20,-100 2`] = `
Object {
  "endX": -330,
  "endY": -50,
  "startX": -440,
  "startY": -200,
}
`;

exports[`start/end/width/height: -20,-100 3`] = `-440`;

exports[`start/end/width/height: -20,-100 4`] = `-200`;

exports[`start/end/width/height: -20,-100 5`] = `110`;

exports[`start/end/width/height: -20,-100 6`] = `150`;

exports[`start/end/width/height: -20,20 1`] = `
Object {
  "endX": -370,
  "endY": -90,
  "startX": -460,
  "startY": -220,
}
`;

exports[`start/end/width/height: -20,20 2`] = `
Object {
  "endX": -370,
  "endY": -90,
  "startX": -460,
  "startY": -220,
}
`;

exports[`start/end/width/height: -20,20 3`] = `-460`;

exports[`start/end/width/height: -20,20 4`] = `-220`;

exports[`start/end/width/height: -20,20 5`] = `90`;

exports[`start/end/width/height: -20,20 6`] = `130`;

exports[`start/end/width/height: -20,60 1`] = `
Object {
  "endX": -370,
  "endY": -90,
  "startX": -480,
  "startY": -160,
}
`;

exports[`start/end/width/height: -20,60 2`] = `
Object {
  "endX": -370,
  "endY": -90,
  "startX": -480,
  "startY": -160,
}
`;

exports[`start/end/width/height: -20,60 3`] = `-480`;

exports[`start/end/width/height: -20,60 4`] = `-160`;

exports[`start/end/width/height: -20,60 5`] = `110`;

exports[`start/end/width/height: -20,60 6`] = `70`;

exports[`start/end/width/height: -60,-20 1`] = `
Object {
  "endX": -270,
  "endY": -110,
  "startX": -360,
  "startY": -120,
}
`;

exports[`start/end/width/height: -60,-20 2`] = `
Object {
  "endX": -270,
  "endY": -110,
  "startX": -360,
  "startY": -120,
}
`;

exports[`start/end/width/height: -60,-20 3`] = `-360`;

exports[`start/end/width/height: -60,-20 4`] = `-120`;

exports[`start/end/width/height: -60,-20 5`] = `90`;

exports[`start/end/width/height: -60,-20 6`] = `10`;

exports[`start/end/width/height: -60,-60 1`] = `
Object {
  "endX": -210,
  "endY": -90,
  "startX": -360,
  "startY": -120,
}
`;

exports[`start/end/width/height: -60,-60 2`] = `
Object {
  "endX": -210,
  "endY": -90,
  "startX": -360,
  "startY": -120,
}
`;

exports[`start/end/width/height: -60,-60 3`] = `-360`;

exports[`start/end/width/height: -60,-60 4`] = `-120`;

exports[`start/end/width/height: -60,-60 5`] = `150`;

exports[`start/end/width/height: -60,-60 6`] = `30`;

exports[`start/end/width/height: -60,-100 1`] = `
Object {
  "endX": -210,
  "endY": -90,
  "startX": -300,
  "startY": -60,
}
`;

exports[`start/end/width/height: -60,-100 2`] = `
Object {
  "endX": -210,
  "endY": -90,
  "startX": -300,
  "startY": -60,
}
`;

exports[`start/end/width/height: -60,-100 3`] = `-300`;

exports[`start/end/width/height: -60,-100 4`] = `-90`;

exports[`start/end/width/height: -60,-100 5`] = `90`;

exports[`start/end/width/height: -60,-100 6`] = `-30`;

exports[`start/end/width/height: -60,20 1`] = `
Object {
  "endX": -270,
  "endY": -110,
  "startX": -420,
  "startY": -100,
}
`;

exports[`start/end/width/height: -60,20 2`] = `
Object {
  "endX": -270,
  "endY": -110,
  "startX": -420,
  "startY": -100,
}
`;

exports[`start/end/width/height: -60,20 3`] = `-420`;

exports[`start/end/width/height: -60,20 4`] = `-110`;

exports[`start/end/width/height: -60,20 5`] = `150`;

exports[`start/end/width/height: -60,20 6`] = `-10`;

exports[`start/end/width/height: -60,60 1`] = `
Object {
  "endX": -330,
  "endY": -50,
  "startX": -420,
  "startY": -100,
}
`;

exports[`start/end/width/height: -60,60 2`] = `
Object {
  "endX": -330,
  "endY": -50,
  "startX": -420,
  "startY": -100,
}
`;

exports[`start/end/width/height: -60,60 3`] = `-420`;

exports[`start/end/width/height: -60,60 4`] = `-100`;

exports[`start/end/width/height: -60,60 5`] = `90`;

exports[`start/end/width/height: -60,60 6`] = `50`;

exports[`start/end/width/height: -100,-20 1`] = `
Object {
  "endX": -50,
  "endY": -10,
  "startX": -200,
  "startY": -120,
}
`;

exports[`start/end/width/height: -100,-20 2`] = `
Object {
  "endX": -50,
  "endY": -10,
  "startX": -200,
  "startY": -120,
}
`;

exports[`start/end/width/height: -100,-20 3`] = `-200`;

exports[`start/end/width/height: -100,-20 4`] = `-120`;

exports[`start/end/width/height: -100,-20 5`] = `150`;

exports[`start/end/width/height: -100,-20 6`] = `110`;

exports[`start/end/width/height: -100,-60 1`] = `
Object {
  "endX": -50,
  "endY": -10,
  "startX": -100,
  "startY": -100,
}
`;

exports[`start/end/width/height: -100,-60 2`] = `
Object {
  "endX": -50,
  "endY": -10,
  "startX": -100,
  "startY": -100,
}
`;

exports[`start/end/width/height: -100,-60 3`] = `-100`;

exports[`start/end/width/height: -100,-60 4`] = `-100`;

exports[`start/end/width/height: -100,-60 5`] = `50`;

exports[`start/end/width/height: -100,-60 6`] = `90`;

exports[`start/end/width/height: -100,-100 1`] = `
Object {
  "endX": 50,
  "endY": 50,
  "startX": -100,
  "startY": -100,
}
`;

exports[`start/end/width/height: -100,-100 2`] = `
Object {
  "endX": 50,
  "endY": 50,
  "startX": -100,
  "startY": -100,
}
`;

exports[`start/end/width/height: -100,-100 3`] = `-100`;

exports[`start/end/width/height: -100,-100 4`] = `-100`;

exports[`start/end/width/height: -100,-100 5`] = `150`;

exports[`start/end/width/height: -100,-100 6`] = `150`;

exports[`start/end/width/height: -100,20 1`] = `
Object {
  "endX": -150,
  "endY": 10,
  "startX": -200,
  "startY": -120,
}
`;

exports[`start/end/width/height: -100,20 2`] = `
Object {
  "endX": -150,
  "endY": 10,
  "startX": -200,
  "startY": -120,
}
`;

exports[`start/end/width/height: -100,20 3`] = `-200`;

exports[`start/end/width/height: -100,20 4`] = `-120`;

exports[`start/end/width/height: -100,20 5`] = `50`;

exports[`start/end/width/height: -100,20 6`] = `130`;

exports[`start/end/width/height: -100,60 1`] = `
Object {
  "endX": -150,
  "endY": 10,
  "startX": -300,
  "startY": -60,
}
`;

exports[`start/end/width/height: -100,60 2`] = `
Object {
  "endX": -150,
  "endY": 10,
  "startX": -300,
  "startY": -60,
}
`;

exports[`start/end/width/height: -100,60 3`] = `-300`;

exports[`start/end/width/height: -100,60 4`] = `-60`;

exports[`start/end/width/height: -100,60 5`] = `150`;

exports[`start/end/width/height: -100,60 6`] = `70`;

exports[`start/end/width/height: 20,-20 1`] = `
Object {
  "endX": -330,
  "endY": -210,
  "startX": -460,
  "startY": -220,
}
`;

exports[`start/end/width/height: 20,-20 2`] = `
Object {
  "endX": -330,
  "endY": -210,
  "startX": -460,
  "startY": -220,
}
`;

exports[`start/end/width/height: 20,-20 3`] = `-460`;

exports[`start/end/width/height: 20,-20 4`] = `-220`;

exports[`start/end/width/height: 20,-20 5`] = `130`;

exports[`start/end/width/height: 20,-20 6`] = `10`;

exports[`start/end/width/height: 20,-60 1`] = `
Object {
  "endX": -350,
  "endY": -190,
  "startX": -460,
  "startY": -220,
}
`;

exports[`start/end/width/height: 20,-60 2`] = `
Object {
  "endX": -350,
  "endY": -190,
  "startX": -460,
  "startY": -220,
}
`;

exports[`start/end/width/height: 20,-60 3`] = `-460`;

exports[`start/end/width/height: 20,-60 4`] = `-220`;

exports[`start/end/width/height: 20,-60 5`] = `110`;

exports[`start/end/width/height: 20,-60 6`] = `30`;

exports[`start/end/width/height: 20,-100 1`] = `
Object {
  "endX": -350,
  "endY": -190,
  "startX": -480,
  "startY": -160,
}
`;

exports[`start/end/width/height: 20,-100 2`] = `
Object {
  "endX": -350,
  "endY": -190,
  "startX": -480,
  "startY": -160,
}
`;

exports[`start/end/width/height: 20,-100 3`] = `-480`;

exports[`start/end/width/height: 20,-100 4`] = `-190`;

exports[`start/end/width/height: 20,-100 5`] = `130`;

exports[`start/end/width/height: 20,-100 6`] = `-30`;

exports[`start/end/width/height: 20,20 1`] = `
Object {
  "endX": -330,
  "endY": -210,
  "startX": -440,
  "startY": -200,
}
`;

exports[`start/end/width/height: 20,20 2`] = `
Object {
  "endX": -330,
  "endY": -210,
  "startX": -440,
  "startY": -200,
}
`;

exports[`start/end/width/height: 20,20 3`] = `-440`;

exports[`start/end/width/height: 20,20 4`] = `-210`;

exports[`start/end/width/height: 20,20 5`] = `110`;

exports[`start/end/width/height: 20,20 6`] = `-10`;

exports[`start/end/width/height: 20,60 1`] = `
Object {
  "endX": -310,
  "endY": -150,
  "startX": -440,
  "startY": -200,
}
`;

exports[`start/end/width/height: 20,60 2`] = `
Object {
  "endX": -310,
  "endY": -150,
  "startX": -440,
  "startY": -200,
}
`;

exports[`start/end/width/height: 20,60 3`] = `-440`;

exports[`start/end/width/height: 20,60 4`] = `-200`;

exports[`start/end/width/height: 20,60 5`] = `130`;

exports[`start/end/width/height: 20,60 6`] = `50`;

exports[`start/end/width/height: 60,-20 1`] = `
Object {
  "endX": -250,
  "endY": -210,
  "startX": -320,
  "startY": -320,
}
`;

exports[`start/end/width/height: 60,-20 2`] = `
Object {
  "endX": -250,
  "endY": -210,
  "startX": -320,
  "startY": -320,
}
`;

exports[`start/end/width/height: 60,-20 3`] = `-320`;

exports[`start/end/width/height: 60,-20 4`] = `-320`;

exports[`start/end/width/height: 60,-20 5`] = `70`;

exports[`start/end/width/height: 60,-20 6`] = `110`;

exports[`start/end/width/height: 60,-60 1`] = `
Object {
  "endX": -250,
  "endY": -210,
  "startX": -380,
  "startY": -300,
}
`;

exports[`start/end/width/height: 60,-60 2`] = `
Object {
  "endX": -250,
  "endY": -210,
  "startX": -380,
  "startY": -300,
}
`;

exports[`start/end/width/height: 60,-60 3`] = `-380`;

exports[`start/end/width/height: 60,-60 4`] = `-300`;

exports[`start/end/width/height: 60,-60 5`] = `130`;

exports[`start/end/width/height: 60,-60 6`] = `90`;

exports[`start/end/width/height: 60,-100 1`] = `
Object {
  "endX": -310,
  "endY": -150,
  "startX": -380,
  "startY": -300,
}
`;

exports[`start/end/width/height: 60,-100 2`] = `
Object {
  "endX": -310,
  "endY": -150,
  "startX": -380,
  "startY": -300,
}
`;

exports[`start/end/width/height: 60,-100 3`] = `-380`;

exports[`start/end/width/height: 60,-100 4`] = `-300`;

exports[`start/end/width/height: 60,-100 5`] = `70`;

exports[`start/end/width/height: 60,-100 6`] = `150`;

exports[`start/end/width/height: 60,20 1`] = `
Object {
  "endX": -190,
  "endY": -190,
  "startX": -320,
  "startY": -320,
}
`;

exports[`start/end/width/height: 60,20 2`] = `
Object {
  "endX": -190,
  "endY": -190,
  "startX": -320,
  "startY": -320,
}
`;

exports[`start/end/width/height: 60,20 3`] = `-320`;

exports[`start/end/width/height: 60,20 4`] = `-320`;

exports[`start/end/width/height: 60,20 5`] = `130`;

exports[`start/end/width/height: 60,20 6`] = `130`;

exports[`start/end/width/height: 60,60 1`] = `
Object {
  "endX": -190,
  "endY": -190,
  "startX": -260,
  "startY": -260,
}
`;

exports[`start/end/width/height: 60,60 2`] = `
Object {
  "endX": -190,
  "endY": -190,
  "startX": -260,
  "startY": -260,
}
`;

exports[`start/end/width/height: 60,60 3`] = `-260`;

exports[`start/end/width/height: 60,60 4`] = `-260`;

exports[`start/end/width/height: 60,60 5`] = `70`;

exports[`start/end/width/height: 60,60 6`] = `70`;
